diff --git a/modules/providers/scald_audio/scald_audio.module b/modules/providers/scald_audio/scald_audio.module
index 93f164f..108eeb8 100644
--- a/modules/providers/scald_audio/scald_audio.module
+++ b/modules/providers/scald_audio/scald_audio.module
@@ -88,14 +88,14 @@ function scald_audio_scald_add_form_fill(&$atoms, $form, $form_state) {
         $atom->title = !empty($ret['tags']['title']) ? $ret['tags']['title'] : $file->filename;
 
         // Prefill the author.
-        $langcode = field_language('scald_atom', $atom, 'scald_authors');
+        $langcode = _scald_atom_form_field_language($atom, 'scald_authors');
         $atom->scald_authors[$langcode][0] = array(
           'tid' => 0,
           'taxonomy_term' => (object)(array('name' => isset($ret['tags']['artist'])?$ret['tags']['artist']:t('Unknown')))
         );
 
         // Prefill tags.
-        $langcode = field_language('scald_atom', $atom, 'scald_tags');
+        $langcode = _scald_atom_form_field_language($atom, 'scald_tags');
         $atom->scald_tags[$langcode][0] = array(
           'tid' => 0,  // Beware, this is not a real tid, it's just an index.
           'taxonomy_term' => (object)(array('name' => isset($ret['tags']['genre'])?$ret['tags']['genre']:t('Unknown')))
@@ -126,7 +126,7 @@ function scald_audio_scald_add_form_fill(&$atoms, $form, $form_state) {
               ->condition('fid', $file->fid)
               ->fields(array('status' => 0))
               ->execute();
-            $langcode = field_language('scald_atom', $atom, 'scald_thumbnail');
+            $langcode = _scald_atom_form_field_language($atom, 'scald_thumbnail');
             $atom->scald_thumbnail[$langcode][0] = (array) $file;
           }
         }
diff --git a/modules/providers/scald_flash/scald_flash.module b/modules/providers/scald_flash/scald_flash.module
index 81927a3..4fd9214 100644
--- a/modules/providers/scald_flash/scald_flash.module
+++ b/modules/providers/scald_flash/scald_flash.module
@@ -54,7 +54,7 @@ function scald_flash_scald_add_form_fill(&$atoms, $form, $form_state) {
     $size_infos = getimagesize($file->uri);
     $atom->data['flash_width'] = $size_infos[0];
     $atom->data['flash_height'] = $size_infos[1];
-    $langcode = field_language('scald_atom', $atom, 'scald_width');
+    $langcode = _scald_atom_form_field_language($atom, 'scald_width');
     $atom->scald_width[$langcode][0]['value'] = $size_infos[0];
     $atom->scald_height[$langcode][0]['value'] = $size_infos[1];
   }
diff --git a/modules/providers/scald_image/scald_image.module b/modules/providers/scald_image/scald_image.module
index 6ee2a9a..5c6a28e 100644
--- a/modules/providers/scald_image/scald_image.module
+++ b/modules/providers/scald_image/scald_image.module
@@ -98,7 +98,6 @@ function scald_image_scald_add_form_fill(&$atoms, $form, $form_state) {
     );
     // Hacky, because variable and field name do not really match.
     foreach ($variable_names as $field_name => $variable_name) {
-      $langcode = field_language('scald_atom', $atom, $field_name);
       if (empty($form_state['values'][$field_name])) {
         continue;
       }
@@ -118,11 +117,13 @@ function scald_image_scald_add_form_fill(&$atoms, $form, $form_state) {
           );
           taxonomy_term_save($term);
         }
+
+        $langcode = _scald_atom_form_field_language($atom, $field_name);
         $atom->{$field_name}[$langcode][] = array('tid' => $term->tid);
       }
     }
 
-    $langcode = field_language('scald_atom', $atom, 'scald_thumbnail');
+    $langcode = _scald_atom_form_field_language($atom, 'scald_thumbnail');
     $atom->scald_thumbnail[$langcode][0] = (array) $file;
   }
 }
diff --git a/scald.module b/scald.module
index 00d6590..3fe98ec 100644
--- a/scald.module
+++ b/scald.module
@@ -433,6 +433,19 @@ function scald_atom_defaults($type) {
   return $defaults[$type] = $default;
 }
 
+/**
+ * Helper for getting the preferred language code for an atom field, useful
+ * when prefilling atom fields in hook_scald_add_form_fill().
+ */
+function _scald_atom_form_field_language($atom, $field_name) {
+  $entity_langcode = entity_language('scald_atom', $atom);
+  $field_langcode = field_valid_language($entity_langcode);
+  $field = field_info_field($field_name);
+  $available_languages = field_available_languages('scald_atom', $field);
+  $languages = _field_language_suggestion($available_languages, $field_langcode, $field_name);
+  return $languages[0];
+}
+
 /*******************************************************************************
  * SCALD ATOM CRUD
  */
